nvflare/lighter/impl/static_file.py

```diff
@@ -220,93 +195,82 @@ def _build_client(self, client, ctx):
+
         # local folder creation
-        dest_dir = self.get_local_dir(client, ctx)
-        utils._write(
-            os.path.join(dest_dir, "log.config.default"),
-            self.template["log_config"],
-            "t",
-        )
-        utils._write(
-            os.path.join(dest_dir, "resources.json.default"),
-            self.template["local_client_resources"],
-            "t",
+        dest_dir = ctx.get_local_dir(client)
+
+        self.build_from_template(ctx, dest_dir, TemplateSectionKey.LOG_CONFIG, ProvFileName.LOG_CONFIG_DEFAULT)
+
+        self.build_from_template(
+            ctx, dest_dir, TemplateSectionKey.LOCAL_CLIENT_RESOURCES, ProvFileName.RESOURCES_JSON_DEFAULT
         )
-        utils._write(
-            os.path.join(dest_dir, "privacy.json.sample"),
-            self.template["sample_privacy"],
-            "t",
+
+        self.build_from_template(
+            ctx,
+            dest_dir,
+            TemplateSectionKey.SAMPLE_PRIVACY,
+            ProvFileName.PRIVACY_JSON_SAMPLE,
         )
-        utils._write(
-            os.path.join(dest_dir, "authorization.json.default"),
-            self.template["default_authz"],
-            "t",
+
+        self.build_from_template(
+            ctx, dest_dir, TemplateSectionKey.DEFAULT_AUTHZ, ProvFileName.AUTHORIZATION_JSON_DEFAULT
         )
 
         # workspace folder file
-        utils._write(
-            os.path.join(self.get_ws_dir(client, ctx), "readme.txt"),
-            self.template["readme_fc"],
-            "t",
-        )
+        dest_dir = ctx.get_ws_dir(client)
+        self.build_from_template(ctx, dest_dir, TemplateSectionKey.CLIENT_README, ProvFileName.README_TXT)
 
-    def _check_host_name(self, host_name: str, server: Participant) -> str:
+    @staticmethod
+    def _check_host_name(host_name: str, server: Participant) -> str:
         if host_name == server.get_default_host():
             # Use the default host - OK
             return ""
 
-        available_host_names = server.get_host_names()
+        available_host_names = server.get_prop(PropKey.HOST_NAMES)
         if available_host_names and host_name in available_host_names:
             # use alternative host name - OK
             return ""
 
         return f"unknown host name '{host_name}'"
 
-    def _prepare_overseer_agent(self, participant, config, role, ctx):
-        project = ctx["project"]
+    def _prepare_overseer_agent(self, participant, config, role, ctx: ProvisionContext):
+        project = ctx.get_project()
         server = project.get_server()
         if not server:
             raise ValueError(f"Missing server definition in project {project.name}")
 
-        fl_port = server.get_prop("fed_learn_port", 8002)
-        admin_port = server.get_prop("admin_port", 8003)
+        fl_port = server.get_prop(PropKey.FED_LEARN_PORT, 8002)
+        admin_port = server.get_prop(PropKey.ADMIN_PORT, 8003)
```

Comment:
should we get the ports using context?

`ctx.get(CtxKey.ADMIN_PORT)`

Reply:
Yes, that's the intent, but I missed this one.